php-frontend/src/main/java/org/sonar/php/parser/PHPGrammar.java

```diff
@@ -372,13 +372,22 @@ public ClassMemberTree CLASS_MEMBER() {
 
   public ClassPropertyDeclarationTree CLASS_CONSTANT_DECLARATION() {
     return b.<ClassPropertyDeclarationTree>nonterminal(PHPLexicalGrammar.CLASS_CONSTANT_DECLARATION).is(
-      f.classConstantDeclaration(
-        b.zeroOrMore(ATTRIBUTE_GROUP()),
-        b.zeroOrMore(CLASS_CONST_MODIFIER()),
-        b.token(PHPKeyword.CONST),
-        MEMBER_CONST_DECLARATION(),
-        b.zeroOrMore(f.newTuple(b.token(COMMA), MEMBER_CONST_DECLARATION())),
-        EOS()));
+      b.firstOf(
+        f.classConstantDeclaration(
+          b.zeroOrMore(ATTRIBUTE_GROUP()),
+          b.zeroOrMore(CLASS_CONST_MODIFIER()),
+          b.token(PHPKeyword.CONST),
+          MEMBER_CONST_DECLARATION(),
+          b.zeroOrMore(f.newTuple(b.token(COMMA), MEMBER_CONST_DECLARATION())),
+          EOS()),
+        f.classConstantDeclaration(
+          b.zeroOrMore(ATTRIBUTE_GROUP()),
+          b.zeroOrMore(CLASS_CONST_MODIFIER()),
+          b.token(PHPKeyword.CONST),
+          DECLARED_TYPE(),
+          MEMBER_CONST_DECLARATION(),
+          b.zeroOrMore(f.newTuple(b.token(COMMA), MEMBER_CONST_DECLARATION())),
+          EOS())));
```

Comment:
Cannot use `b.optional(...)` for the `DECLARED_TYPE` -> for example, in `const A`, it would match the `A` as the `DECLARED_TYPE`, then fail and not be able to consider that there is no `DECLARED_TYPE`, only solution I found is to do two separated matching and a `b.firstOf(...)` call.

Reply:
Strange that it seems to be working with `CLASS_VARIABLE_DECLARATION`, but I also can't figure out, why.

Reply:
For me, this is coming from the fact that for variable declaration, their name starts with `$`, which is not possible for a `DECLARED_TYPE`, so it breaks any ambiguity and cannot match the variable name for a possible `DECLARED_TYPE`.
Const does not have a `$` in their name.
```
private string $a
private const string A
```

Reply:
Unfortunately, this will be the solution I would also prefer. It's not perfect, but the other solution would be to make the decision in the factory which I don't prefer.